pkg/runner/collector_selector.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+package runner
+
+import (
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/redhat-partner-solutions/vse-sync-testsuite/pkg/collectors"
+)
+
+var (
+	optionalCollectorNames []string
+	requiredCollectorNames []string
+	All                    string = "all"
+)
+
+func init() {
+	optionalCollectorNames = make([]string, 0)
+	optionalCollectorNames = append(optionalCollectorNames,
+		collectors.DPLLCollectorName,
+		collectors.GPSCollectorName,
+	)
+
+	requiredCollectorNames = make([]string, 0)
+	requiredCollectorNames = append(requiredCollectorNames,
+		collectors.DevInfoCollectorName,
+	)
+}
+
+func isIn(name string, arr []string) bool {
+	for _, arrVal := range arr {
+		if name == arrVal {
+			return true
+		}
+	}
+	return false
+}
+
+func removeDuplicates(arr []string) []string {
+	res := make([]string, 0)
+	for _, name := range arr {
+		if !isIn(name, res) {
+			res = append(res, name)
+		}
+	}
+	return res
+}
+
+func GetCollectorsNames(selectedCollectors []string) []string {
+	collectorNames := make([]string, 0)
+	for _, name := range selectedCollectors {
+		switch {
+		case strings.EqualFold(name, "all"):
+			collectorNames = append(collectorNames, requiredCollectorNames...)
+			collectorNames = append(collectorNames, optionalCollectorNames...)
+			collectorNames = removeDuplicates(collectorNames)
+			return collectorNames
+		case isIn(name, collectorNames):
+			continue
+		case isIn(name, requiredCollectorNames), isIn(name, optionalCollectorNames):
+			collectorNames = append(collectorNames, name)
+		default:
+			log.Errorf("Unknown collector %s. Ignored", name)
+		}
+	}
+	missingCollectors := make([]string, 0)
+	for _, requiredName := range requiredCollectorNames {
+		if !isIn(requiredName, collectorNames) {
+			missingCollectors = append(missingCollectors, requiredName)
+		}
+	}
+	if len(missingCollectors) > 0 {
+		log.Warnf(
+			"The following required collectors were missing %s. They will be added",
```

Comment:
I think this method can be significantly simplified. If this warning is replaced by a message (which could also be handled by the caller) simply stating the final collectors that will be run.

That would also enable initialising `collectorNames` as a copy of `requiredCollectorNames` then only appending, which would simplify the whole `switch` statement.

e.g. instead of building a slice of missing required collectors just output something like `fmt.Sprintf("Running Collectors: %s", strings.Join(collectorNames, ","))`.

Alternatively building the slice of `missingCollectors` could be replaced with a comparison of the length of `requiredCollectorNames` and `collectorNames`, if they are different then warn that required ones have been added.